File: Revision/Sorting/mergeSort.cpp
#include<bits/stdc++.h>
using namespace std;
void merge(int arr[],int start,int mid,int end){
vector<int> temp;
int i=start;
int j=mid+1;
while(i<=mid && j<=end){
    if(arr[i]<=arr[j]){
        temp.push_back(arr[i]);
        i++;
    }else{
        temp.push_back(arr[j]);
        j++;
    }
}

while(i<=mid){
 temp.push_back(arr[i]);
        i++;
}
while(j<=end){
    temp.push_back(arr[j]);
    j++;
}

for(int i=start;i<=end;i++){
    arr[i]=temp[i-start];
}




}
void mergeSort(int arr[],int start,int end){
    if(start>=end) return;
    int mid = start+((end-start)/2);
    mergeSort(arr,start,mid);
    mergeSort(arr,mid+1,end);
    merge(arr,start,mid,end);  
}
int main(){

    int arr[]={7,8,9,1,2,3,6,5,4};
    int sizes = sizeof(arr)/sizeof(arr[0]);
    mergeSort(arr,0,sizes-1);
    for(auto it:arr){
        cout<<it<<" ";
    }
    
}

File: JS/Sorting/mergeSort.js
let arr=[9,8,7,6,5,4,3,2,1];

function merge(arr,start,mid,end){
let i=start;
let j = mid+1;
let sample=[]
while(i<=mid && j<=end){
    if(arr[i]<=arr[j]){
        sample.push(arr[i])
        i++;
    }else{
        sample.push(arr[j])
        j++;
    }
}
while(i<=mid){
    sample.push(arr[i])
    i++;
}
while(j<=end)
{
    sample.push(arr[j])
    j++;
}
for(let a = 0;a<sample.length;a++){
    arr[start+a]=sample[a]
}


}
function mergeSort(arr,start,end){
if(start>=end){
    return
}
let mid = Math.floor(start+((end-start)/2))
mergeSort(arr,start,mid)
mergeSort(arr,mid+1,end)
merge(arr,start,mid,end)


}
mergeSort(arr,0,arr.length-1)

for(a of arr){
    process.stdout.write(`${a} `)
}